### PR TITLE
Adding feature CharSequence.mismatches

### PR DESCRIPTION
**Pull request by shardulsonar**
----
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/master/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

File: apis/cc-en_GB/atrium-api-cc-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/cc/en_GB/charSequenceAssertions.kt [hunk 1]
```diff
@@ -5,6 +5,7 @@ import ch.tutteli.atrium.api.cc.en_GB.creating.charsequence.contains.builders.No
 import ch.tutteli.atrium.api.cc.en_GB.creating.charsequence.contains.builders.impl.NotCheckerOptionImpl
 import ch.tutteli.atrium.creating.Assert
 import ch.tutteli.atrium.creating.AssertionPlant
+import ch.tutteli.atrium.creating.Expect
```
Comment (robstoll) [inline apis/cc-en_GB/atrium-api-cc-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/cc/en_GB/charSequenceAssertions.kt]: please remove the import
```suggestion

```

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/charSequenceAssertions.kt [hunk 2]
```diff
@@ -3,6 +3,7 @@ package ch.tutteli.atrium.api.fluent.en_GB
 import ch.tutteli.atrium.api.fluent.en_GB.creating.charsequence.contains.builders.NotCheckerOption
 import ch.tutteli.atrium.api.fluent.en_GB.creating.charsequence.contains.builders.impl.NotCheckerOptionImpl
 import ch.tutteli.atrium.creating.Expect
+import ch.tutteli.atrium.domain.builders.AssertImpl
```
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/charSequenceAssertions.kt]: please remove the import
```suggestion

```

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/charSequenceAssertions.kt [hunk 3]
```diff
@@ -197,3 +198,12 @@ fun <T : CharSequence> Expect<T>.isNotEmpty() = addAssertion(ExpectImpl.charSequ
...
 fun <T : CharSequence> Expect<T>.isNotBlank() = addAssertion(ExpectImpl.charSequence.isNotBlank(this))
+
+/**
+ * Expects that the subject of the assertion (a [CharSequence]) mis matches given [Regex].
+ *
+ * @return This assertion container to support a fluent API.
+ * @throws AssertionError Might throw an [AssertionError] if the assertion made is not correct.
+ */
+fun <T : CharSequence> Expect<T>.mismatches(expected: Regex) =
+    addAssertion(AssertImpl.charSequence.mismatches(this, expected))
```
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/charSequenceAssertions.kt]:
```suggestion
    addAssertion(ExpectImpl.charSequence.mismatches(this, expected))
```

File: domain/robstoll-lib/atrium-domain-robstoll-lib-common/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/charSequenceAssertions.kt [hunk 4]
```diff
@@ -41,3 +42,6 @@ fun _isNotEmpty(subjectProvider: SubjectProvider<CharSequence>): Assertion =
 
 fun _isNotBlank(subjectProvider: SubjectProvider<CharSequence>): Assertion =
     AssertImpl.builder.createDescriptive(subjectProvider, IS_NOT, RawString.create(BLANK)) { it.isNotBlank() }
+
+fun <T : CharSequence> _mismatches(assertionContainer: Expect<T>, expected: Regex): Assertion =
+    AssertImpl.builder.createDescriptive(assertionContainer, MISMATCHES, expected) {!it.matches(expected)}
```
Comment (robstoll) [inline domain/robstoll-lib/atrium-domain-robstoll-lib-common/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/charSequenceAssertions.kt]: Please use always ExpectImpl
```suggestion
    ExpectImpl.builder.createDescriptive(assertionContainer, MISMATCHES, expected) { !it.matches(expected) }
```

File: misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/CharSequenceAssertionsSpec.kt [hunk 5]
```diff
@@ -212,4 +214,25 @@ abstract class CharSequenceAssertionsSpec(
             }
         }
     }
+
+    describeFun(mismatches.name){
+        // TODO : matchesFun
```
Comment (robstoll) [inline misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/CharSequenceAssertionsSpec.kt]:
```suggestion

```

File: misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/CharSequenceAssertionsSpec.kt [hunk 6]
```diff
@@ -212,4 +214,25 @@ abstract class CharSequenceAssertionsSpec(
...
+        val mismatchesFun = mismatches.lambda
+
+        context("text '$text"){
+            // TODO : Matches Negative Condition
+
```
Comment (robstoll) [inline misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/CharSequenceAssertionsSpec.kt]:
```suggestion

```

File: misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/CharSequenceAssertionsSpec.kt [hunk 7]
```diff
@@ -212,4 +214,25 @@ abstract class CharSequenceAssertionsSpec(
...
+            it("${mismatches.name} 'Hello' does not throw"){
+                fluent.mismatchesFun(Regex("Hello"))
+            }
+
+            // TODO : Matches Negative Condition
+
```
Comment (robstoll) [inline misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/CharSequenceAssertionsSpec.kt]:
```suggestion

```

**Comment by robstoll** [comment]
@shardulsonar Looks almost good, there are only small subtle things you need to correct (you can also `commit suggestion` via UI)

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/charSequenceAssertions.kt [hunk 8]
```diff
@@ -197,3 +197,12 @@ fun <T : CharSequence> Expect<T>.isNotEmpty() = addAssertion(ExpectImpl.charSequ
  * @throws AssertionError Might throw an [AssertionError] if the assertion made is not correct.
  */
 fun <T : CharSequence> Expect<T>.isNotBlank() = addAssertion(ExpectImpl.charSequence.isNotBlank(this))
+
+/**
+ * Expects that the subject of the assertion (a [CharSequence]) mis matches given [Regex].
```
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/charSequenceAssertions.kt]:
```suggestion
 * Expects that the subject of the assertion (a [CharSequence]) mismatches given [Regex].
```

File: misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/CharSequenceAssertionsSpec.kt [hunk 9]
```diff
@@ -212,4 +214,22 @@ abstract class CharSequenceAssertionsSpec(
             }
         }
     }
+
+    describeFun(mismatches.name){
+        val mismatchesFun = mismatches.lambda
+
+        context("text '$text"){
+            // TODO : Matches Negative Condition
```
Comment (robstoll) [inline misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/CharSequenceAssertionsSpec.kt]: is there a reason why you keep the TODO? (same question applies to the TODO below)

Comment (shardulsonar) [reply misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/CharSequenceAssertionsSpec.kt]: I kept it for better understanding of the flow. I was not supposed push that. I thought I removed all the TODO statements. But I forgot. Fixed now

**Comment by robstoll** [comment]
@shardulsonar thanks for the contribution :+1:
Fancy another one? https://github.com/robstoll/atrium/issues?q=is%3Aissue+is%3Aopen+label%3Ahacktoberfest+no%3Aassignee

**Comment by shardulsonar** [comment]
> @shardulsonar thanks for the contribution 👍
> Fancy another one? https://github.com/robstoll/atrium/issues?q=is%3Aissue+is%3Aopen+label%3Ahacktoberfest+no%3Aassignee

Definately yes. I am not available for a week. I will pick up issues once I am back next week.

**Pull request merged**